local-cluster/tests/local_cluster.rs

```diff
@@ -5506,3 +5490,156 @@ fn test_duplicate_shreds_switch_failure() {
+                break;
+            }
+        }
+
+        assert!(
+            timer.elapsed() < Duration::from_secs(30),
+            "Did not make more than 10 blocks in 30 seconds"
+        );
+        std::thread::yield_now();
+    }
+
+    // Send duplicate
+    let parent = {
+        let blockstore = open_blockstore(&target_ledger_path);
+        let parent = blockstore
+            .meta(dup_slot)
+            .unwrap()
+            .unwrap()
+            .parent_slot
+            .unwrap();
+
+        let entries = create_ticks(
+            64 * (std::cmp::max(1, dup_slot - parent)),
+            0,
+            cluster.genesis_config.hash(),
+        );
+        let last_hash = entries.last().unwrap().hash;
+        let version = solana_sdk::shred_version::version_from_hash(&last_hash);
+        let dup_shreds = Shredder::new(dup_slot, parent, 0, version)
+            .unwrap()
+            .entries_to_shreds(
+                &majority_keypair,
+                &entries,
+                true,  // is_full_slot
+                0,     // next_shred_index,
+                0,     // next_code_index
+                false, // merkle_variant
+                &ReedSolomonCache::default(),
+                &mut ProcessShredsStats::default(),
+            )
+            .0;
+
+        info!("Sending duplicate shreds for {dup_slot}");
+        cluster.send_shreds_to_validator(dup_shreds.iter().collect(), &target_pubkey);
+        wait_for_duplicate_proof(&target_ledger_path, dup_slot)
+            .expect("Duplicate proof for {dup_slot} not found");
+        parent
+    };
+
+    info!("Duplicate proof for {dup_slot} has landed, restarting node");
+    let info = cluster.exit_node(&target_pubkey);
+
+    {
+        let blockstore = open_blockstore(&target_ledger_path);
+        purge_slots_with_count(&blockstore, dup_slot + 5, 100);
+    }
+
+    // Restart, should create an entirely new fork
+    cluster.restart_node(&target_pubkey, info, SocketAddrSpace::Unspecified);
+
+    info!("Waiting for fork built off {parent}");
+    let timer = Instant::now();
+    let mut checked_children: HashSet<Slot> = HashSet::default();
+    let mut done = false;
+    while !done {
+        let blockstore = open_blockstore(&target_ledger_path);
+        let parent_meta = blockstore.meta(parent).unwrap().expect("Meta must exist");
+        for child in parent_meta.next_slots {
+            if checked_children.contains(&child) {
+                continue;
+            }
+
+            if blockstore.is_full(child) {
+                let shreds = blockstore
+                    .get_data_shreds_for_slot(child, 0)
+                    .expect("Child is full");
+                let mut our_block = true;
+                for shred in shreds {
+                    our_block &= shred.verify(&target_pubkey);
+                }
+                if our_block {
+                    done = true;
+                }
+                checked_children.insert(child);
```

Comment:
We still might build a block off the duplicate block instead of the parent in the new proposed change to deal with the votes not landing right?

Would it be better to check that our next vote is not on the duplicate fork heaviest (have to guarantee that the duplicate fork is the heaviest though, i.e. votes have landed on it)?

Reply:
Even with the new proposed change, if you have an invalid fork with no alternative valid for you will always create a new fork, which is the case here since we are only 1 fork. So although after we create the new fork we might go back to building off the duplicate fork, since we check all children this will still work. 

>Would it be better to check that our next vote is not on the duplicate fork heaviest (have to guarantee that the duplicate fork is the heaviest though, i.e. votes have landed on it)?

This could be possible although it would require a really complicated setup, because as is the validator would not be able to ever switch onto the non duplicate fork, so we would need a switching threshold stake to both be not on the duplicate fork but also not have created their own fork already. 
I'm don't think this setup would test the intended code path in a more rigorous way, and is probably better for a duplicate switch rollback test.